Add prompts when installer aborts terminal or needs install location

lechacon opened the pull request:
Resolves: #713
Resolves: #970 

Changes:
* Added user setting for default install root
* Added prompts when an installer has `InstallerAbortsTerminal` (to confirm that we want to install) or `InstallLocationRequired` (to provide the install location).
* Refactored the existing package agreements prompt to a common setup also used by the two new prompts. Before installing a package, we walk through all the possible prompts and show them if applicable. When installing multiple packages, we check all the packages that need the prompt and then show the prompt once.

Remaining items:
* Testing
* Use install root from settings
* Update settings doc
* Move installers that abort the terminal to the end on upgrade all/import
* Skip prompts when running non-interactively

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/1814)

JohnMcPMS reviewed (changes requested):
My comments on the default install root setting also apply to the install location argument on a multipackage install.  I'm just not sure it makes sense to have one location apply to multiple packages.

lechacon commented on src/AppInstallerCLICore/ExecutionReporter.cpp:
```diff
@@ -155,6 +155,27 @@ namespace AppInstaller::CLI::Execution
         }
     }
 
+    std::filesystem::path Reporter::PromptForPath(Resource::LocString message, Level level)
```
Re 1. I'm adding a Context flag to set non-interactive mode. The flag is set when constructing a COMContext and can be added any other place we need it. Then each prompt has to handle the flag however it needs to (e.g. the one for abort terminal does nothing, and the one for agreements throws an error).

Re 2 & 3. @denelon, what do you think about adding an argument and setting to suppress prompts (agreements, missing install location, warning about installer aborting the terminal)?

JohnMcPMS replied:
RE 1: Agreed that each location will have to decide what the non-interactive default is, and then some or all of them will need to have additional arguments to select the non-default behavior.

RE 2&3: I don't think this is an option, but a requirement.

JohnMcPMS replied:
Also, you cannot have the agreement throwing an error in the COM case without:

1. Exposing the agreements through COM
2. Enabling the ability to accept the agreements through COM
3. Giving a grace period for callers to start using this pattern

lechacon replied:
Added a context flag for COM, a CLI arg and a setting to indicate non-interactive.

I'm not sure what to do for a proper handling of the agreements via COM, but this change does not leave us worse off than we currently are. Currently, when a COM context is created, a flag is set that signals that the agreements were accepted by the caller, so it will not throw.

lechacon commented on doc/Settings.md:
````diff
@@ -90,6 +90,8 @@ The `architectures` behavior affects what architectures will be selected when in
     },
 ```
 
+### Default install root
````
Added details

lechacon commented on schemas/JSON/settings/settings.schema.0.2.json:
```diff
@@ -79,7 +79,12 @@
         "preferences": { "$ref": "#/definitions/InstallPrefReq" },
-        "requirements": { "$ref": "#/definitions/InstallPrefReq" }
+        "requirements": { "$ref": "#/definitions/InstallPrefReq" },
+        "defaultInstallRoot": {
```
Updated to always append the package ID to the root (option 2)

lechacon commented:
Note that this is still missing tests, validation of the install root and the arg/setting to disable prompts. (Hence still a draft)

I had thought of moving installers that abort the terminal to the end so as to minimize their impact, but that seems tricky with dependencies.

lechacon commented:
Reviving this old PR...

Main changes from last time:
* Rebased on recent main branch
* Added a setting and a CLI arg to prevent all prompts
* Added unit tests

JohnMcPMS commented on src/AppInstallerCLICore/Argument.cpp:
```diff
@@ -110,6 +110,7 @@ namespace AppInstaller::CLI
         args.push_back(ForType(Args::Type::RainbowStyle));
         args.push_back(ForType(Args::Type::RetroStyle));
         args.push_back(ForType(Args::Type::VerboseLogs));
+        args.emplace_back("disable-interactivity", NoAlias, Args::Type::DisableInteractivity, Resource::String::DisableInteractivityArgumentDescription, ArgumentType::Standard, false);
```
Should this be a flag?  You use it like one in PromptFlow, so I think yes.

lechacon replied:
You're right. Changed to flag.

JohnMcPMS commented on src/AppInstallerCLICore/Workflows/SourceFlow.cpp:
```diff
@@ -3,6 +3,7 @@
 #include "pch.h"
 #include "Resources.h"
 #include "SourceFlow.h"
+#include "PromptFlow.h"
```
Seems like a leftover from a previous design?

lechacon replied:
No, this is needed. The tasks for prompting for source agreements are used here; they were in WorkflowBase but I moved them to PromptFlow to have all the handling of interactivity in one place.

lechacon commented on schemas/JSON/settings/settings.schema.0.2.json:
```diff
@@ -116,6 +116,11 @@
+        "defaultInstallRoot": {
+          "description": "Default install location to use for packages that require it when not specified",
+          "type": "string",
+          "maxLength": "1024"
```
Changed to 32,767

lechacon commented on doc/Settings.md:
````diff
@@ -183,6 +193,20 @@ The `doProgressTimeoutInSeconds` setting updates the number of seconds to wait w
+
+```json
+    "interactivity": {
+        "disable": true
````
I agree, but I'm going to leave it as "disable" to match the one for "telemetry" and the `--disable-interactivity` flag. I think `disable: true` is easy to understand, and I don't imagine the confusing `disable: false` to be that common in this case. If this turns out to be a bad idea, we have plenty of time before next release to change it.

lechacon commented:
This PR is having issues with the COM InProc E2E tests because I added something that causes us to load the string resources in a place we weren't doing it before, and the tests fail there because they cannot find the resources. I still don't know how to fix it (and it would make this change bigger), so I'm sidestepping it for now by preventing this new use of the ResourceLoader. I'll open an issue for actually fixing it because I imagine it will hit us again later.
